File: sunyardlib/src/main/java/com/netplus/sunyardlib/ReceiptBuilder.java
package com.netplus.sunyardlib;

import android.util.Log;

import com.netpluspay.terminal_core.AndroidTerminalReceiptBuilderFactory;
import com.socsi.smartposapi.printer.Align;
import com.socsi.smartposapi.printer.FontLattice;
import com.socsi.smartposapi.printer.FontType;
import com.socsi.smartposapi.printer.PrintRespCode;
import com.socsi.smartposapi.printer.Printer2;
import com.socsi.smartposapi.printer.TextEntity;

import io.reactivex.Single;

/**
 * ReceiptBuilder
 */
public class ReceiptBuilder extends AndroidTerminalReceiptBuilderFactory<ReceiptBuilder, Single<PrintRespCode>> {

    private final FontType mCh = FontType.SIMSUM;
    private final FontType mEn = FontType.SIMSUM;
    private final Printer2 printer;
    StringBuilder builder = new StringBuilder();
    public ReceiptBuilder() {
        this.printer = Printer2.getInstance();
    }

    @Override
    protected ReceiptBuilder getThis() {
        return this;
    }

    @Override
    public Single<PrintRespCode> print() {
        build();
        Log.e("TAG", builder.toString());
        PrintRespCode respCode = printer.startPrint();
        return Single.just(respCode);
    }

    @Override
    protected void appendTextEntity(String str) {
        builder.append(str);
        printer.appendTextEntity2(new TextEntity(str, mCh, mEn, FontLattice.TWENTY_FOUR, false, Align.LEFT, true));
    }

    @Override
    protected void appendTextEntityBold(String str) {
        builder.append(str);
        printer.appendTextEntity2(new TextEntity(str, mCh, mEn, FontLattice.TWENTY_FOUR, true, Align.LEFT, true));
    }

    @Override
    protected void appendTextEntityFontSixteen(String str) {
        builder.append(str);
        printer.appendTextEntity2(new TextEntity(str, mCh, mEn, FontLattice.FORTY_EIGHT, true, Align.LEFT, true));
    }

    @Override
    protected void appendTextEntityFontSixteenCenter(String str) {
        builder.append(str);
        printer.appendTextEntity2(new TextEntity(str, mCh, mEn, FontLattice.FORTY_EIGHT, true, Align.CENTER, true));
    }

    @Override
    protected void appendTextEntityCenter(String str) {
        builder.append(str);
        printer.appendTextEntity2(new TextEntity(str, mCh, mEn, FontLattice.TWENTY_FOUR, false, Align.CENTER, true));
    }
}

File: kozenlib/src/main/java/com/netpluspay/kozenlib/printer/ReceiptBuilder.kt
package com.netpluspay.kozenlib.printer

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.netpluspay.kozenlib.KozenLib
import com.netpluspay.kozenlib.R
import com.netpluspay.terminalcore.AndroidTerminalReceiptBuilderFactory
import com.pos.sdk.printer.IPosPrinterListener
import com.pos.sdk.printer.POIPrinterManage
import com.pos.sdk.printer.models.BitmapPrintLine
import com.pos.sdk.printer.models.PrintLine
import com.pos.sdk.printer.models.TextPrintLine
import io.reactivex.Single


class ReceiptBuilder :
    AndroidTerminalReceiptBuilderFactory<ReceiptBuilder, Single<PrinterResponse>>() {

    private val textPrintLine = TextPrintLine().apply {
        type = PrintLine.TEXT
    }
    private val printerManager = KozenLib.getPrinterManager().apply {
        setPrintGray(Integer.valueOf("4000"))
        setLineSpace(Integer.valueOf("1"))
        cleanCache()
    }

    override fun getThis(): ReceiptBuilder = this


    override fun appendTextEntity(p0: String?) {
        printerManager.addPrintLine(
            textPrintLine.apply {
                content = p0
                position = PrintLine.LEFT
                size = TextPrintLine.FONT_NORMAL
                isBold = false
                isItalic = false
                isInvert = false
            })
    }

    override fun appendTextEntityBold(p0: String?) {
        printerManager.addPrintLine(
            textPrintLine.apply {
                content = p0
                position = PrintLine.LEFT
                size = TextPrintLine.FONT_NORMAL
                isBold = true
                isItalic = false
                isInvert = false
            })
    }

    override fun appendTextEntityFontSixteen(p0: String?) {
        printerManager.addPrintLine(
            textPrintLine.apply {
                content = p0
                position = PrintLine.LEFT
                size = TextPrintLine.FONT_NORMAL
                isBold = false
                isItalic = false
                isInvert = false
            })
    }

    override fun appendTextEntityFontSixteenCenter(p0: String?) {
        printerManager.addPrintLine(
            textPrintLine.apply {
                content = p0
                position = PrintLine.CENTER
                size = TextPrintLine.FONT_LARGE
                isBold = true
                isItalic = false
                isInvert = false
            })
    }

    override fun appendTextEntityCenter(p0: String?) {
        printerManager.addPrintLine(
            textPrintLine.apply {
                content = p0
                position = PrintLine.CENTER
                size = TextPrintLine.FONT_NORMAL
                isBold = false
                isItalic = false
                isInvert = false
            })
    }

    fun print(printerListener: POIPrinterManage.IPrinterListener){
        build()
        printerManager.beginPrint(printerListener)
    }

    override fun print(): Single<PrinterResponse> =
        Single.create {
            build()
            printerManager.beginPrint(object : POIPrinterManage.IPrinterListener {
                override fun onError(p0: Int, p1: String?) {
                    it.onError(Throwable("message:$p1 - code:$p0"))
                }

                override fun onFinish() {
                    it.onSuccess(PrinterResponse())
                }

                override fun onStart() {
                    Log.e("TAG", "Printing started")
                }
            })
        }

    override fun appendImageCenter() {
        val bitmapPrintLine = BitmapPrintLine()
        bitmapPrintLine.type = PrintLine.BITMAP
        bitmapPrintLine.position = PrintLine.CENTER
        val bitmap: Bitmap = BitmapFactory.decodeResource(KozenLib.getContext().resources, R.drawable.ic_netpos_logo)
        bitmapPrintLine.bitmap = Bitmap.createScaledBitmap(bitmap, 120, 120, false)
        printerManager.addPrintLine(bitmapPrintLine)
    }
}
